### PR TITLE
Implement instruction builder for transfering SOL or SPL tokens

### PR DESCRIPTION
**Pull request by yrliou**
Resolves https://github.com/brave/brave-browser/issues/20946

Reference impl for transferring SOL:
https://docs.rs/solana-program/latest/src/solana_program/system_instruction.rs.html#390
https://github.com/solana-labs/solana-web3.js/blob/62513cdee2ca0c4f350e811a982ca13731201a9f/src/system-program.ts#L694-L699

Reference impl for transferring SPL tokens
https://docs.rs/spl-token/3.3.0/src/spl_token/instruction.rs.html#890
https://github.com/solana-labs/solana-program-library/blob/8eaaafed54d3fd4978e5d2fa0f2ee7f07dc9a0a7/token/js/client/token.js#L1511

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/browser/solana_instruction_builder.cc [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+                               {SolanaAccountMeta(from_pubkey, true, true),
+                                SolanaAccountMeta(to_pubkey, false, true)}),
+                           instruction_data);
+}
+
+}  // namespace system_program
+
+namespace spl_token_program {
+
+absl::optional<SolanaInstruction> Transfer(
```
Comment (yrliou) [inline components/brave_wallet/browser/solana_instruction_builder.cc:73]: See https://docs.rs/spl-token/latest/src/spl_token/instruction.rs.html#105 for the interface.

File: components/brave_wallet/browser/solana_instruction_builder.cc [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+}  // namespace
+
+namespace brave_wallet {
+
+namespace solana {
+
+namespace system_program {
+
+absl::optional<SolanaInstruction> Transfer(const std::string& from_pubkey,
```
Comment (yrliou) [inline components/brave_wallet/browser/solana_instruction_builder.cc:46]: See https://docs.rs/solana-program/1.9.5/src/solana_program/system_instruction.rs.html#174 for the interface.

File: components/brave_wallet/browser/solana_instruction_builder.cc [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+#include "brave/components/brave_wallet/browser/brave_wallet_constants.h"
+#include "brave/components/brave_wallet/browser/solana_account_meta.h"
+#include "brave/components/brave_wallet/browser/solana_instruction.h"
+#include "brave/components/brave_wallet/common/brave_wallet_types.h"
+#include "build/build_config.h"
+
+namespace {
+
+template <typename T>
+void UintToLEBytes(T val, std::vector<uint8_t>* bytes) {
```
Comment (darkdh) [inline components/brave_wallet/browser/solana_instruction_builder.cc:19]: It's better to put comment saying this is doing `bincode::serialize` to encode number

Comment (yrliou) [reply components/brave_wallet/browser/solana_instruction_builder.cc:21]: Addressed in https://github.com/brave/brave-core/pull/12170/commits/6fb9d4d34cee63614dc57f708700f8b36365379e

File: components/brave_wallet/browser/brave_wallet_constants.h [hunk 4]
```diff
@@ -619,6 +619,7 @@ constexpr int kSolanaBlockhashSize = 32;
 // 1232 = 1280(IPv6 minimum MTU) - 40(size of the IPv6 header) - 8(size of the
 // fragment header)
 constexpr size_t kSolanaMaxTxSize = 1232;
+constexpr char kSolanaSystemProgramId[] = "11111111111111111111111111111111";
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/browser/brave_wallet_constants.h:622]: Addressed in https://github.com/brave/brave-core/pull/12170/commits/6fb9d4d34cee63614dc57f708700f8b36365379e

**Review by darkdh — APPROVED** [review]

**Pull request merged**